src/main/java/com/uber/cadence/internal/sync/WorkflowRetryerInternal.java

```diff
@@ -88,14 +94,19 @@ public static void retry(RetryOptions options, Functions.Proc proc) {
    * @return result promise to the result or failure if retries stopped according to options.
    */
   public static <R> Promise<R> retryAsync(RetryOptions options, Functions.Func<Promise<R>> func) {
-    long startTime = Workflow.currentTimeMillis();
-    return retryAsync(options, func, startTime, 1);
+    String retryId = WorkflowInternal.randomUUID().toString();
+    long startTime = WorkflowInternal.currentTimeMillis();
+    return retryAsync(retryId, options, func, startTime, 1);
   }
 
   private static <R> Promise<R> retryAsync(
-      RetryOptions options, Functions.Func<Promise<R>> func, long startTime, long attempt) {
+      String retryId,
+      RetryOptions options,
+      Functions.Func<Promise<R>> func,
+      long startTime,
+      long attempt) {
     options.validate();
-    CompletablePromise<R> funcResult = Workflow.newPromise();
+    CompletablePromise<R> funcResult = WorkflowInternal.newCompletablePromise();
```

Comment:
We need to call WorkflowInternal.mutableSideEffect here to record retry option on the first (non-replay) attempt.